### DISCUSSION TITLE
Help writing a Clickhouse query to display percentages of total over time in Grafana

### DISCUSSION THREAD

**Discussion by cjesper** [ClickHouse/ClickHouse#45902] [Q&A]
Hello, 
Apologies if questions like these are better placed elsewhere - this was the best fit i could fit

I've recently started using ClickHouse for a use case that influxdb (which i'm used to) was unable to handle due to the large volumes/cardinality involved. It's been working great but there is a specific view that i've been unable to re-create in clickhouse/SQL. I am relatively new to SQL as well.

Basically i have a table that represents the time required to perform certain actions in my program:
Time(ms) | Action 
150000.     Action A
223131.      Action B
15000        Action C
etc etc. 
I want to compute what % of the total time spent each action is responsible for. So:
% Of total | Action
85.              Action A
10               Action B
5                 Action C

Doing this in normal SQL/without trying to consider time is quite trivial:
```
SELECT action, (sum(sum) * 100 / (
      SELECT sum(sum) FROM my_table WHERE $__timeFilter(time))) as percentage
      FROM my_table
WHERE $__timeFilter(time) 
GROUP BY action
```

The above query gives me a table containing the percentage of the total time each action has taken the last **timeFilter** hours
However, i also want to see this grouped by time (hourly) - that is, it should compute the above for every hour. This has given me quite a lot of trouble. I have tried various approaches - here is one:

```
SELECT
    time,
    action,
    action_sum,
    total_sum,
    round((action_sum / total_sum) * 100, 2) AS perc
FROM
(
    SELECT
        toStartOfTenMinutes(time) AS time,
        action,
        sum(sum) AS action_sum,
        (
            SELECT sum(sum)
            FROM my_table
            WHERE  $__timeFilter(time)
        ) AS total_sum
    FROM my_table
    WHERE  $__timeFilter(time)
    GROUP BY
        time,
        action
)
ORDER BY
    time ASC,
```

This query gives me an action_sum, total_sum, and percentage grouped by time (10 minutes) but:
1. The percentages are off - they don't add up to 100% which indicates something is wrong.
2. I can only visualize the result in a table. If i try to view it as a time series in Grafana I get this:
![image](https://user-images.githubusercontent.com/15783742/216089086-656a0848-12c6-4872-bc83-171208ec7646.png)
ie. it loses the per-action grouping. I am not sure if this is a problem with the grafana visualisation, my query, or both.

Any help would be greatly appreciated :)

**Comment by mdonkers** [comment]
I have limited knowledge about Grafana, so won't be able to provide too much help. Perhaps our blog might give some pointers: https://clickhouse.com/blog/visualizing-data-with-grafana
From the query I think the problem is only single values are returned. So like in the top query, also in the latter one you probably need to `GROUP BY action` so that you get separate sets of data.

And with the values not adding up, seems like the issue is in the `total_sum` calculation? You're grouping by time and action, but the `total_sum` is fetched for the full time range. This should be the total over all actions, but within the time bucket.

Not sure if this can be done with a normal GROUP BY, perhaps better to use a WINDOW function (see for example: https://clickhouse.com/docs/en/sql-reference/window-functions/#maximumtotal-salary-per-department)
Something like:
```sql
SELECT
    action,
    time,
    total_sum_per_wndw,
    round((sum / total_sum_per_wndw) * 100, 2) AS `share_per_wndw(%)`
FROM
(
    SELECT
        action,
        toStartOfTenMinutes(time) AS time,
        sum,
        sum(sum) OVER wndw AS total_sum_per_wndw
    FROM my_table
    WINDOW wndw AS (PARTITION BY toStartOfTenMinutes(time)
          rows BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
    ORDER BY
        time ASC
);

```
